File: harnessarena_uploader/parsers/opencode.py
from __future__ import annotations

import json
import sqlite3
from datetime import datetime, timezone
from pathlib import Path
from typing import Optional

from ..base_parser import HarnessParser
from ..history_paths import get_opencode_history_paths
from ..metric_strategies import HarnessMetricStrategies
from ..helpers import _basename_only, _make_session_id, _parse_timestamp, _register_mcp_tool, _safe_int
from ..models import Harness, SessionMeta, TokenUsage, ToolCallSummary
from ..trackers import TimeSpanTracker, PlanModeTracker


class OpenCodeParser(HarnessParser):
    """OpenCode session parser.

    OpenCode stores sessions in SQLite (Drizzle ORM). Tool calls appear in the
    `part` table with type in (tool-call, tool_use, function_call) and toolName.
    """

    harness_type = Harness.OPENCODE

    def __init__(self) -> None:
        self._paths = get_opencode_history_paths()
        self._strategies = HarnessMetricStrategies.snapshot_defaults()

    def parse(self, since: Optional[datetime] = None) -> list[SessionMeta]:
        return _parse_opencode(since, parser=self, paths=self._paths)

    def metric_strategies(self) -> HarnessMetricStrategies:
        return self._strategies


def _parse_opencode(
    since: Optional[datetime] = None,
    parser: Optional[HarnessParser] = None,
    paths=None,
) -> list[SessionMeta]:
    """Parse OpenCode SQLite database.

    Path: ~/.local/share/opencode/opencode.db
    Tables:
      session: id, project_id, title, directory, time_created
      message: id, session_id, data JSON with role/model/tokens/cost
      part: id, message_id, data JSON with type/text
    """
    paths = paths or get_opencode_history_paths()
    db_path = paths.db_path
    if not db_path.is_file():
        return []

    results: list[SessionMeta] = []

    try:
        conn = sqlite3.connect(f"file:{db_path}?mode=ro", uri=True)
        conn.row_factory = sqlite3.Row
        cursor = conn.cursor()

        # time_created is Unix epoch milliseconds in OpenCode
        query = "SELECT id, project_id, parent_id, title, directory, time_created, permission FROM session"
        params: list = []
        if since:
            query += " WHERE time_created >= ?"
            params.append(int(since.replace(tzinfo=timezone.utc).timestamp() * 1000))

        cursor.execute(query, params)
        sessions = cursor.fetchall()

        # Count subagent spawns per parent session
        # Sessions with non-null parent_id are subagent sessions
        spawn_counts: dict[str, int] = {}
        for s in sessions:
            parent = s["parent_id"]
            if parent:
                spawn_counts[str(parent)] = spawn_counts.get(str(parent), 0) + 1

        for session_row in sessions:
            session_id = str(session_row["id"])
            parent_id = str(session_row["parent_id"]) if session_row["parent_id"] else None
            project_name = _basename_only(session_row["directory"])
            started_at = _parse_timestamp(session_row["time_created"])

            # Detect plan mode from permission field
            # OpenCode stores permissions as JSON array with plan_enter/plan_exit entries
            # action="allow" means plan mode is enabled for the session
            plan_entries = 0
            try:
                perms = json.loads(session_row["permission"]) if session_row["permission"] else []
                if isinstance(perms, list):
                    for perm in perms:
                        if isinstance(perm, dict) and perm.get("permission") == "plan_enter":
                            if perm.get("action") == "allow":
                                plan_entries = 1
            except (json.JSONDecodeError, TypeError):
                pass

            # Aggregate message metadata — read only role, model ID, tokens, cost
            # NEVER read content, summary text, or diffs
            cursor.execute(
                "SELECT data FROM message WHERE session_id = ?",
                (session_row["id"],),
            )

            user_count = 0
            assistant_count = 0
            total_count = 0
            input_tokens = 0
            output_tokens = 0
            cost = 0.0
            model = "unknown"
            provider: Optional[str] = None
            ended_at = ""
            # Track agent/mode from message data for plan mode + skill detection
            agent_names: set[str] = set()
            skill_invocations: dict[str, int] = {}

            for msg_row in cursor.fetchall():
                total_count += 1
                try:
                    data = json.loads(msg_row["data"]) if isinstance(msg_row["data"], str) else {}
                except json.JSONDecodeError:
                    continue

                role = data.get("role", "")
                if role == "user":
                    user_count += 1
                elif role == "assistant":
                    assistant_count += 1

                # Track agent name for plan mode and subagent detection
                # "build" = normal, "plan" = plan mode, other = custom agent (not a skill)
                agent_name = data.get("agent", "")
                if agent_name and agent_name not in ("build", ""):
                    agent_names.add(agent_name)

                # Model can be at data.modelID or data.model.modelID
                m = data.get("modelID") or data.get("model")
                if isinstance(m, dict):
                    m = m.get("modelID")
                if m:
                    model = m

                # Provider
                p = data.get("providerID")
                if p:
                    provider = p

                # Timestamps for end time
                time_info = data.get("time", {}) or {}
                completed = time_info.get("completed")
                if completed:
                    ended_at = _parse_timestamp(completed)

                tokens = data.get("tokens", {}) or {}
                input_tokens += _safe_int(tokens.get("input"))
                output_tokens += _safe_int(tokens.get("output"))

                c = data.get("cost")
                if c is not None:
                    try:
                        cost += float(c)
                    except (TypeError, ValueError):
                        pass

            # Build time spans from message timestamps
            time_tracker = TimeSpanTracker()
            plan_tracker = PlanModeTracker()
            _in_plan_mode = False
            for msg_row2 in cursor.execute(
                "SELECT time_created, data FROM message WHERE session_id = ? ORDER BY time_created",
                (session_row["id"],),
            ).fetchall():
                try:
                    mdata = json.loads(msg_row2["data"]) if isinstance(msg_row2["data"], str) else {}
                except json.JSONDecodeError:
                    continue
                msg_dt = datetime.fromtimestamp(msg_row2["time_created"] / 1000, tz=timezone.utc)
                msg_role = mdata.get("role", "")
                msg_agent = mdata.get("agent", "")
                if msg_role == "user":
                    time_tracker.on_user_turn(msg_dt)
                elif msg_role == "assistant":
                    time_tracker.on_nonuser_event(msg_dt)
                # Track plan mode transitions by agent field changes
                if msg_agent == "plan" and not _in_plan_mode:
                    _in_plan_mode = True
                    plan_tracker.on_enter(msg_dt)
                elif msg_agent != "plan" and _in_plan_mode:
                    _in_plan_mode = False
                    plan_tracker.on_exit(msg_dt)

            if _in_plan_mode and time_tracker._last_nonuser_ts:
                plan_tracker.on_exit(time_tracker._last_nonuser_ts)

            time_spans, turn_exec_times = time_tracker.finalize()
            # Add plan mode spans
            for ps in plan_tracker.spans:
                time_spans.append(ps)
            # Replace harness_exec with plan_mode for session-level plan
            plan_tracker.replace_session_level(
                time_spans, max(plan_entries, int("plan" in agent_names))
            )

            # Count tool-type parts and extract tool spans
            cursor.execute(
                """SELECT p.data FROM part p
                   JOIN message m ON p.message_id = m.id
                   WHERE m.session_id = ?""",
                (session_row["id"],),
            )
            tool_call_count = 0
            tool_counts: dict[str, int] = {}
            tool_categories: dict[str, str] = {}
            subagent_calls = 0
            background_agents = 0
            mcp_calls = 0
            plan_mode_entries = 0
            plan_mode_exits = 0
            mcp_servers: dict[str, dict] = {}
            for part_row in cursor.fetchall():
                try:
                    pdata = json.loads(part_row["data"]) if isinstance(part_row["data"], str) else {}
                except json.JSONDecodeError:
                    continue
                ptype = pdata.get("type", "")
                # OpenCode uses type="tool" with tool name in "tool" field
                # Other formats use tool-call/tool_use/function_call with toolName/name
                is_tool = False
                tool_name = "unknown_tool"
                if ptype in ("tool-call", "tool_use", "function_call"):
                    tool_name = pdata.get("toolName", pdata.get("name", "unknown_tool"))
                    is_tool = True
                elif ptype == "tool" and "tool" in pdata:
                    tool_name = pdata["tool"]
                    is_tool = True
                # Detect skill invocations: tool="skill" with state.input.name
                if is_tool and tool_name == "skill":
                    state = pdata.get("state", {})
                    if isinstance(state, dict):
                        meta = state.get("metadata", {}) or {}
                        skill_name = (state.get("input", {}) or {}).get("name") or meta.get("name", "")
                        if skill_name:
                            # Determine source from metadata.dir path
                            skill_dir = meta.get("dir", "")
                            if ".claude/skills/" in skill_dir:
                                source = "user-custom"
                            elif "/.opencode/" in skill_dir or "plugins" in skill_dir:
                                source = "marketplace"
                            else:
                                source = "user-custom"
                            skill_invocations[skill_name] = {"count": skill_invocations.get(skill_name, {}).get("count", 0) + 1, "source": source}

                if is_tool:
                    tool_counts[tool_name] = tool_counts.get(tool_name, 0) + 1
                    tool_call_count += 1
                    # Classify tool category
                    if tool_name == "skill":
                        tool_categories[tool_name] = "skill"
                    elif tool_name == "task":
                        tool_categories[tool_name] = "subagent"
                    elif tool_name.startswith("mcp_") or tool_name.startswith("mcp__"):
                        tool_categories[tool_name] = "mcp"
                    else:
                        tool_categories.setdefault(tool_name, "tool")
                    # Extract tool span timing from state.time
                    state = pdata.get("state", {})
                    if isinstance(state, dict):
                        time_info = state.get("time", {})
                        if isinstance(time_info, dict) and time_info.get("start") and time_info.get("end"):
                            try:
                                t_start = datetime.fromtimestamp(time_info["start"] / 1000, tz=timezone.utc)
                                t_end = datetime.fromtimestamp(time_info["end"] / 1000, tz=timezone.utc)
                                dur = (t_end - t_start).total_seconds()
                                if dur >= 0:
                                    # Find which harness_exec span this tool falls in and add to its tool_spans
                                    for sp in time_spans:
                                        if sp.get("type") == "harness_exec" or sp.get("type") == "plan_mode":
                                            sp_start = datetime.fromisoformat(sp["start"])
                                            sp_end = datetime.fromisoformat(sp["end"])
                                            if sp_start <= t_start <= sp_end:
                                                if "tool_spans" not in sp:
                                                    sp["tool_spans"] = []
                                                sp["tool_spans"].append({
                                                    "name": tool_name,
                                                    "category": "tool",
                                                    "start": t_start.isoformat(),
                                                    "end": t_end.isoformat(),
                                                    "seconds": round(dur, 3),
                                                })
                                                break
                            except (TypeError, ValueError, OSError):
                                pass
                    if parser:
                        tool_input = pdata.get("args", pdata.get("input", {}))
                        # OpenCode nests input in state.input
                        if not isinstance(tool_input, dict):
                            state = pdata.get("state", {})
                            tool_input = state.get("input", {}) if isinstance(state, dict) else {}
                        if not isinstance(tool_input, dict):
                            tool_input = {}
                        c = parser.classify_tool_call(tool_name, tool_input)
                        if c["is_subagent"]:
                            subagent_calls += 1
                            if c["is_background_agent"]:
                                background_agents += 1
                        if c["is_mcp"]:
                            mcp_calls += 1
                            _register_mcp_tool(mcp_servers, tool_name)
                        if c["is_plan_enter"]:
                            plan_mode_entries += 1
                        if c["is_plan_exit"]:
                            plan_mode_exits += 1

            if total_count == 0:
                continue

            # Duration
            duration = None
            if started_at and ended_at:
                try:
                    s = datetime.fromisoformat(started_at.replace("Z", "+00:00"))
                    e = datetime.fromisoformat(ended_at.replace("Z", "+00:00"))
                    duration = max(0, int((e - s).total_seconds()))
                except ValueError:
                    pass

            results.append(parser.session_from_snapshot({
                "id": _make_session_id(Harness.OPENCODE, session_id),
                "source_session_id": session_id,
                "parent_session_id": parent_id,
                "agent_name": next((a for a in agent_names if a != "plan"), None) if parent_id else None,
                "harness_version": None,
                "project_name": project_name,
                "git_repo_name": project_name,
                "git_branch": None,
                "model": model,
                "provider": provider,
                "message_count_user": user_count,
                "message_count_assistant": assistant_count,
                "message_count_total": total_count,
                "tool_call_count": tool_call_count,
                "tokens": TokenUsage(
                    input_tokens=input_tokens,
                    output_tokens=output_tokens,
                    total_tokens=input_tokens + output_tokens,
                ),
                "subagent_calls": max(subagent_calls, spawn_counts.get(session_id, 0)),
                "background_agents": background_agents,
                "mcp_calls": mcp_calls,
                "mcp_servers": mcp_servers,
                "plan_mode_entries": max(plan_entries, plan_mode_entries, int("plan" in agent_names)),
                "plan_mode_exits": max(plan_entries, plan_mode_exits, int("plan" in agent_names)),
                "tool_calls": tuple(
                    ToolCallSummary(name, count, tool_categories.get(name, "tool"))
                    for name, count in sorted(tool_counts.items())
                ),
                "skills_used": skill_invocations,
                "cost_usd": cost if cost > 0 else None,
                "started_at": started_at,
                "ended_at": ended_at if ended_at else None,
                "duration_seconds": duration,
                "time_spans": time_spans,
                "total_exec_seconds": round(sum(t for t in turn_exec_times), 1) if turn_exec_times else None,
                "mean_turn_seconds": round(sum(turn_exec_times) / len(turn_exec_times), 1) if turn_exec_times else None,
                "median_turn_seconds": round(sorted(turn_exec_times)[len(turn_exec_times) // 2], 1) if turn_exec_times else None,
            }))

        conn.close()
    except sqlite3.Error:
        pass

    return results

File: src/parsers/opencode.ts
import * as fs from "node:fs";

import Database from "better-sqlite3";

import { HarnessParser } from "../base-parser.js";
import { basenameOnly, makeSessionId, parseTimestamp, registerMcpTool, safeInt } from "../helpers.js";
import {
  getOpenCodeHistoryPaths,
  type OpenCodeHistoryPaths,
} from "../history-paths.js";
import { snapshotDefaults, type HarnessMetricStrategies } from "../metric-strategies.js";
import {
  createTokenUsage,
  Harness,
  type SessionMeta,
  type ToolCallSummary,
} from "../models.js";
import { PlanModeTracker, TimeSpanTracker, type TimeSpan, type ToolSpan } from "../trackers.js";

export class OpenCodeParser extends HarnessParser {
  readonly harnessType = Harness.OPENCODE;

  private readonly _paths: OpenCodeHistoryPaths;
  private readonly _strategies: HarnessMetricStrategies;

  constructor() {
    super();
    this._paths = getOpenCodeHistoryPaths();
    this._strategies = snapshotDefaults();
  }

  parse(since?: Date): SessionMeta[] {
    return parseOpenCode(since, this, this._paths);
  }

  metricStrategies(): HarnessMetricStrategies {
    return this._strategies;
  }
}

function parseOpenCode(
  since: Date | undefined,
  parser: OpenCodeParser,
  paths: OpenCodeHistoryPaths,
): SessionMeta[] {
  const dbPath = paths.dbPath;
  if (!fs.existsSync(dbPath)) return [];

  const results: SessionMeta[] = [];

  try {
    const db = new Database(dbPath, { readonly: true });

    let query =
      "SELECT id, project_id, parent_id, title, directory, time_created, permission FROM session";
    const params: unknown[] = [];
    if (since) {
      query += " WHERE time_created >= ?";
      params.push(Math.floor(since.getTime()));
    }

    const sessions = db.prepare(query).all(...params) as Record<string, unknown>[];

    // Count subagent spawns per parent session
    const spawnCounts = new Map<string, number>();
    for (const s of sessions) {
      const parent = s.parent_id;
      if (parent) {
        const parentStr = String(parent);
        spawnCounts.set(parentStr, (spawnCounts.get(parentStr) ?? 0) + 1);
      }
    }

    for (const sessionRow of sessions) {
      const sessionId = String(sessionRow.id);
      const parentId = sessionRow.parent_id ? String(sessionRow.parent_id) : null;
      const projectName = basenameOnly(sessionRow.directory as string | null);
      const startedAt = parseTimestamp(sessionRow.time_created);

      // Detect plan mode from permission field
      let planEntries = 0;
      try {
        const perms = sessionRow.permission
          ? JSON.parse(sessionRow.permission as string)
          : [];
        if (Array.isArray(perms)) {
          for (const perm of perms) {
            if (
              typeof perm === "object" &&
              perm !== null &&
              (perm as Record<string, unknown>).permission === "plan_enter" &&
              (perm as Record<string, unknown>).action === "allow"
            ) {
              planEntries = 1;
            }
          }
        }
      } catch {
        // ignore
      }

      // Aggregate message metadata
      const msgRows = db
        .prepare("SELECT data FROM message WHERE session_id = ?")
        .all(sessionRow.id) as { data: unknown }[];

      let userCount = 0;
      let assistantCount = 0;
      let totalCount = 0;
      let inputTokens = 0;
      let outputTokens = 0;
      let cost = 0;
      let model = "unknown";
      let provider: string | null = null;
      let endedAt = "";
      const agentNames = new Set<string>();
      const skillInvocations: Record<string, Record<string, unknown>> = {};

      for (const msgRow of msgRows) {
        totalCount++;
        let data: Record<string, unknown>;
        try {
          data =
            typeof msgRow.data === "string" ? JSON.parse(msgRow.data) : {};
        } catch {
          continue;
        }

        const role = (data.role as string) ?? "";
        if (role === "user") userCount++;
        else if (role === "assistant") assistantCount++;

        // Track agent name for plan mode and subagent detection
        const agentName = (data.agent as string) ?? "";
        if (agentName && agentName !== "build" && agentName !== "") {
          agentNames.add(agentName);
        }

        // Model can be at data.modelID or data.model.modelID
        let m: unknown = data.modelID ?? data.model;
        if (typeof m === "object" && m !== null) {
          m = (m as Record<string, unknown>).modelID;
        }
        if (m) model = String(m);

        // Provider
        const p = data.providerID;
        if (p) provider = String(p);

        // Timestamps for end time
        const timeInfo = (data.time ?? {}) as Record<string, unknown>;
        const completed = timeInfo.completed;
        if (completed) endedAt = parseTimestamp(completed);

        const tokens = (data.tokens ?? {}) as Record<string, unknown>;
        inputTokens += safeInt(tokens.input);
        outputTokens += safeInt(tokens.output);

        const c = data.cost;
        if (c != null) {
          const costVal = Number(c);
          if (!Number.isNaN(costVal)) cost += costVal;
        }
      }

      // Build time spans from message timestamps
      const timeTracker = new TimeSpanTracker();
      const planTracker = new PlanModeTracker();
      let inPlanMode = false;

      const orderedMsgRows = db
        .prepare(
          "SELECT time_created, data FROM message WHERE session_id = ? ORDER BY time_created",
        )
        .all(sessionRow.id) as { time_created: number; data: unknown }[];

      for (const msgRow2 of orderedMsgRows) {
        let mdata: Record<string, unknown>;
        try {
          mdata =
            typeof msgRow2.data === "string" ? JSON.parse(msgRow2.data) : {};
        } catch {
          continue;
        }
        const msgDt = new Date(msgRow2.time_created);
        const msgRole = (mdata.role as string) ?? "";
        const msgAgent = (mdata.agent as string) ?? "";
        if (msgRole === "user") {
          timeTracker.onUserTurn(msgDt);
        } else if (msgRole === "assistant") {
          timeTracker.onNonuserEvent(msgDt);
        }
        // Track plan mode transitions by agent field changes
        if (msgAgent === "plan" && !inPlanMode) {
          inPlanMode = true;
          planTracker.onEnter(msgDt);
        } else if (msgAgent !== "plan" && inPlanMode) {
          inPlanMode = false;
          planTracker.onExit(msgDt);
        }
      }

      if (inPlanMode && timeTracker.lastNonuserTimestamp) {
        planTracker.onExit(timeTracker.lastNonuserTimestamp);
      }

      const { timeSpans, turnExecTimes } = timeTracker.finalize();
      // Add plan mode spans
      for (const ps of planTracker.spans) {
        timeSpans.push(ps);
      }
      // Replace harness_exec with plan_mode for session-level plan
      planTracker.replaceSessionLevel(
        timeSpans,
        Math.max(planEntries, agentNames.has("plan") ? 1 : 0),
      );

      // Count tool-type parts and extract tool spans
      const partRows = db
        .prepare(
          `SELECT p.data FROM part p
           JOIN message m ON p.message_id = m.id
           WHERE m.session_id = ?`,
        )
        .all(sessionRow.id) as { data: unknown }[];

      let toolCallCount = 0;
      const toolCounts = new Map<string, number>();
      const toolCategories = new Map<string, string>();
      let subagentCalls = 0;
      let backgroundAgents = 0;
      let mcpCalls = 0;
      let planModeEntries = 0;
      let planModeExits = 0;
      const mcpServers: Record<string, Record<string, unknown>> = {};

      for (const partRow of partRows) {
        let pdata: Record<string, unknown>;
        try {
          pdata =
            typeof partRow.data === "string" ? JSON.parse(partRow.data) : {};
        } catch {
          continue;
        }

        const ptype = (pdata.type as string) ?? "";
        let isTool = false;
        let toolName = "unknown_tool";

        if (ptype === "tool-call" || ptype === "tool_use" || ptype === "function_call") {
          toolName = (pdata.toolName as string) ?? (pdata.name as string) ?? "unknown_tool";
          isTool = true;
        } else if (ptype === "tool" && "tool" in pdata) {
          toolName = pdata.tool as string;
          isTool = true;
        }

        // Detect skill invocations: tool="skill" with state.input.name
        if (isTool && toolName === "skill") {
          const state = pdata.state as Record<string, unknown> | undefined;
          if (typeof state === "object" && state !== null) {
            const meta = (state.metadata ?? {}) as Record<string, unknown>;
            const stateInput = (state.input ?? {}) as Record<string, unknown>;
            const skillName =
              (stateInput.name as string) ?? (meta.name as string) ?? "";
            if (skillName) {
              const skillDir = (meta.dir as string) ?? "";
              let source: string;
              if (skillDir.includes(".claude/skills/")) {
                source = "user-custom";
              } else if (
                skillDir.includes("/.opencode/") ||
                skillDir.includes("plugins")
              ) {
                source = "marketplace";
              } else {
                source = "user-custom";
              }
              const existing = (skillInvocations[skillName] ?? {
                count: 0,
                source,
              }) as Record<string, unknown>;
              skillInvocations[skillName] = {
                count: ((existing.count as number) ?? 0) + 1,
                source,
              };
            }
          }
        }

        if (isTool) {
          toolCounts.set(toolName, (toolCounts.get(toolName) ?? 0) + 1);
          toolCallCount++;

          // Classify tool category
          if (toolName === "skill") {
            toolCategories.set(toolName, "skill");
          } else if (toolName === "task") {
            toolCategories.set(toolName, "subagent");
          } else if (toolName.startsWith("mcp_") || toolName.startsWith("mcp__")) {
            toolCategories.set(toolName, "mcp");
          } else if (!toolCategories.has(toolName)) {
            toolCategories.set(toolName, "tool");
          }

          // Extract tool span timing from state.time
          const state = pdata.state as Record<string, unknown> | undefined;
          if (typeof state === "object" && state !== null) {
            const timeInfo = state.time as Record<string, unknown> | undefined;
            if (
              typeof timeInfo === "object" &&
              timeInfo !== null &&
              timeInfo.start &&
              timeInfo.end
            ) {
              try {
                const tStart = new Date(Number(timeInfo.start));
                const tEnd = new Date(Number(timeInfo.end));
                const dur = (tEnd.getTime() - tStart.getTime()) / 1000;
                if (dur >= 0) {
                  // Find which harness_exec span this tool falls in
                  for (const sp of timeSpans) {
                    if (sp.type === "harness_exec" || sp.type === "plan_mode") {
                      const spStart = new Date(sp.start);
                      const spEnd = new Date(sp.end);
                      if (spStart <= tStart && tStart <= spEnd) {
                        if (!sp.tool_spans) sp.tool_spans = [];
                        (sp.tool_spans as ToolSpan[]).push({
                          name: toolName,
                          category: "tool",
                          start: tStart.toISOString(),
                          end: tEnd.toISOString(),
                          seconds: Math.round(dur * 1000) / 1000,
                        });
                        break;
                      }
                    }
                  }
                }
              } catch {
                // ignore
              }
            }
          }

          let toolInput = (pdata.args ?? pdata.input ?? {}) as Record<string, unknown>;
          // OpenCode nests input in state.input
          if (typeof toolInput !== "object" || toolInput === null) {
            const stateObj = pdata.state as Record<string, unknown> | undefined;
            toolInput =
              typeof stateObj === "object" && stateObj !== null
                ? ((stateObj.input ?? {}) as Record<string, unknown>)
                : {};
          }
          if (typeof toolInput !== "object" || toolInput === null) toolInput = {};

          const c = parser.classifyToolCall(toolName, toolInput);
          if (c.is_subagent) {
            subagentCalls++;
            if (c.is_background_agent) backgroundAgents++;
          }
          if (c.is_mcp) {
            mcpCalls++;
            registerMcpTool(mcpServers, toolName);
          }
          if (c.is_plan_enter) planModeEntries++;
          if (c.is_plan_exit) planModeExits++;
        }
      }

      if (totalCount === 0) continue;

      // Duration
      let duration: number | null = null;
      if (startedAt && endedAt) {
        try {
          const s = new Date(startedAt.replace("Z", "+00:00"));
          const e = new Date(endedAt.replace("Z", "+00:00"));
          duration = Math.max(0, Math.floor((e.getTime() - s.getTime()) / 1000));
        } catch {
          // ignore
        }
      }

      const toolCallSummaries: ToolCallSummary[] = [...toolCounts.entries()]
        .sort(([a], [b]) => a.localeCompare(b))
        .map(([name, count]) => ({
          tool_name: name,
          invocation_count: count,
          category: toolCategories.get(name) ?? "tool",
        }));

      results.push(
        parser.sessionFromSnapshot({
          id: makeSessionId(Harness.OPENCODE, sessionId),
          source_session_id: sessionId,
          parent_session_id: parentId,
          agent_name: parentId
            ? [...agentNames].find((a) => a !== "plan") ?? null
            : null,
          harness_version: null,
          project_name: projectName,
          git_repo_name: projectName,
          git_branch: null,
          model,
          provider,
          message_count_user: userCount,
          message_count_assistant: assistantCount,
          message_count_total: totalCount,
          tool_call_count: toolCallCount,
          tokens: createTokenUsage({
            input_tokens: inputTokens,
            output_tokens: outputTokens,
            total_tokens: inputTokens + outputTokens,
          }),
          subagent_calls: Math.max(subagentCalls, spawnCounts.get(sessionId) ?? 0),
          background_agents: backgroundAgents,
          mcp_calls: mcpCalls,
          mcp_servers: mcpServers,
          plan_mode_entries: Math.max(
            planEntries,
            planModeEntries,
            agentNames.has("plan") ? 1 : 0,
          ),
          plan_mode_exits: Math.max(
            planEntries,
            planModeExits,
            agentNames.has("plan") ? 1 : 0,
          ),
          tool_calls: toolCallSummaries,
          skills_used: skillInvocations,
          cost_usd: cost > 0 ? cost : null,
          started_at: startedAt,
          ended_at: endedAt || null,
          duration_seconds: duration,
          time_spans: timeSpans,
          total_exec_seconds:
            turnExecTimes.length > 0
              ? Math.round(turnExecTimes.reduce((a, b) => a + b, 0) * 10) / 10
              : null,
          mean_turn_seconds:
            turnExecTimes.length > 0
              ? Math.round(
                  (turnExecTimes.reduce((a, b) => a + b, 0) / turnExecTimes.length) * 10,
                ) / 10
              : null,
          median_turn_seconds:
            turnExecTimes.length > 0
              ? Math.round(
                  [...turnExecTimes].sort((a, b) => a - b)[
                    Math.floor(turnExecTimes.length / 2)
                  ] * 10,
                ) / 10
              : null,
        }),
      );
    }

    db.close();
  } catch {
    // ignore sqlite errors
  }

  return results;
}
